packages/vm/src/imports.rs

```diff
@@ -150,19 +151,24 @@ macro_rules! write_region {
 }
 
 /// Reads a storage entry from the VM's storage into Wasm memory
-pub fn do_read<S: Storage, Q: Querier>(
-    ctx: &mut Ctx,
-    key_ptr: u32,
-    value_ptr: u32,
-) -> VmResult<i32> {
+pub fn do_read<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32) -> VmResult<i32> {
     let key = read_region!(ctx, key_ptr, MAX_LENGTH_DB_KEY);
     // `Ok(expr?)` used to convert the error variant.
     let value: Option<Vec<u8>> =
         with_storage_from_context::<S, Q, _, _>(ctx, |store| Ok(store.get(&key)?))?;
-    Ok(match value {
-        Some(buf) => write_region!(ctx, value_ptr, &buf),
-        None => errors::read::KEY_DOES_NOT_EXIST,
-    })
+
+    let out_data = match value {
+        Some(data) => data,
+        None => return Ok(errors::read::KEY_DOES_NOT_EXIST),
+    };
+
+    let out_ptr = with_func_from_context::<S, Q, u32, u32, _, _>(ctx, "allocate", |allocate| {
+        let out_size = to_u32(out_data.len())?;
+        let ptr = allocate.call(out_size)?;
+        Ok(ptr)
+    })?;
+    write_region!(ctx, out_ptr, &out_data);
+    Ok(to_i32(out_ptr)?)
 }
```

Comment:
As discussed privately, after these changes:
* All the errors here and in `do_write` are not errors that need to be reported to the user, except missing values in storage, so:
  * `write_region!(...)` can be replaced with `write_region(...)?`
  * `read_region!(...)` can be replaced with `read_region(...)?`
  * instead of returning `KEY_DOES_NOT_EXIST` we can just return `0` which can be interpreted as a null pointer, in contract with the `Ok(to_i32(out_ptr)?)` in the end of the function which returns the pointer to a value which was found in storage.
* In `cosmwasm_std` we can simplify `Storage` and `ReadonlyStorage` even further to not return `Result`s at all.

Both of the points above can be done as future PRs.

Reply:
Getting rid of the error cases is actually necessary to represent the 4 GB address space of Wasm correctly in a `u32` return type. It turns out that [in Wasm, i32 is not the same as in Rust](https://webassembly.github.io/spec/core/syntax/values.html#syntax-int). This explains why we are able to use both Rust's `u32` and `i32` over the same import interface.

This `u32` return type of `db_read` with 0 being _key does not exist_ should be covered in this PR.

Reply:
oof
this also impacts all the other APIs the contract consumes...

Reply:
But in no other place we mix error code and pointer. So it should be trivial to update the rest.

It would even be easier if we get a guarantee that those pointers are an offset in the memory and all memory usage up to 2 GB always is in the signed 32 bit range. But I don't know this.

Reply:
It's probably better to go on the safe side and not mix pointers and errors, as you said.